android/src/main/java/com/robinpowered/RNFrequency/RNFrequencyModule.java

```diff
@@ -0,0 +1,119 @@
+package com.robinpowered.RNFrequency;
+
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+import com.facebook.react.bridge.ReactMethod;
+import com.facebook.react.bridge.Promise;
+
+import android.media.AudioFormat;
+import android.media.AudioManager;
+import android.media.AudioTrack;
+import android.os.Build;
+
+public class RNFrequencyModule extends ReactContextBaseJavaModule {
+    private static final String MODULE_NAME = "FrequencyManager";
+    private static final String TRACK_STOPPED_PLAYING = "TRACK_STOPPED_PLAYING";
+    private static final int SAMPLE_RATE = 44100;
+
+    private AudioTrack track;
+    private Promise promise;
+
+    public RNFrequencyModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+    }
+
+    private void stop() {
+        // If track is initialized, stop playback and release resources
+        try {
+            track.stop();
+        } catch (IllegalStateException e) {
+            // no-op
+        }
+
+        track.release();
+        track = null;
+
+        if (promise !=  null) {
+            promise.reject(TRACK_STOPPED_PLAYING, "Track stopped playing");
+            promise = null;
+        }
+    }
+
+    private void saveAudioTrackAndPromise(AudioTrack audioTrack, Promise promise) {
+        track = audioTrack;
+        this.promise = promise;
```

Comment:
This is what I was referring to by creating a tuple

Rather than having two fields for the track + promise, have a class that owns the two of them so they can be managed together.

Example of what we can clean this class up to look like:

```java
private void cancel() {
  audioSession.cancel()
  audioSession = null
}

private void complete() {
  audioSession.complete()
  audioSession = null
}

@ReactMethod
public void playFrequency(...) {
  if (audioSession && audioSession.isPlaying()) {
    stop()
  }

  // separate logic for creating the track
  AudioTrack track = createAudioTrack(frequency, duration)
  audioSession = new AudioSession(track, promise)

  track.setNotificationMarkerPosition(..)
  track.setPlaybackPositionUpdateListener(new AudioTrack.OnPlaybackPositionUpdateListener() {
    ...
    public void onMarkerReached(AudioTrack track) {
      FrequencyModule.this.complete()
    })
  })

  track.play()
}

// this manages the coordination of events that need to be shared between
// the track and the promise
class AudioSession {
  Promise promise;
  AudioTrack track

  cancel() {
    track.stop()
    track.release()
    promise.reject(..)
  }

  complete() {
    track.stop()
    track.release()
    promise.resolve(..)
  }

  isPlaying() {
    return track.getState() != ... && track.etPlayState() !== ...
  }
}
```

Reply:
Just like what we were talking about last week!

As I had suggested back then, this class could be an extension of the AudioTrack class, wrapping it and just adding additional functionality of resolving the promise. 

`class AudioSession extends AudioTrack {`
You can override the `play` and `stop` class methods, resolving or rejecting the promises accordingly, creating your own lifecycle inside the class.

Reply:
Composition > inheritance in this situation. We're not creating a special version of an AudioTrack, we're creating a facade that coordinates the calls between two different things.